BGP-CP: Remove NodeSpecer and ControlPlaneState

ldelossa opened the pull request:
This PR removes the NodeSpecer abstraction and the ControlPlaneState structure from the BGP control plane. 

Now that LocalNodeStore is available it can be used in place of these two constructs. 

Additionally, since all ConfigReconcilers are now part of the Hive/Cell dependency injection graph, each ConfigReconciler can declare their own depedencies at time of construction. This provides much greater flexibility as a ConfigReconciler can now utilize any entities available in the DIG to perform their necessary reconciliation. 

```release-note
Remove NodeSpecer and ControlPlaneState from BGP-CP. Rely on Hive/Cell for further ConfigReconciler dependencies. 
```

ldelossa committed:
pkg/node: expose 'newTestLocalNodeStore' for testing

This commit makes the previously private 'newTestLocalNodeStore'
constructor public and also updates the constructor to receive a mock
LocalNode structure.

The returned 'LocalNodeStore' will then "emit" the mock LocalNode when
invoked.

This is useful for unit tests which include the 'LocalNodeStore'
structure as a depedency, avoiding the need to create a Hive for the
unit test.

These changes will be used in subsequent commits.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
bgpv1,annotations: introduce ResolveRouterID method

Introduce a helper method for resolving the RouterID with a parsed
BGP-CP annotation.

This will be used in subsequent commits

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
bgpv1: use LocalNodeStore to eliminiate ControlPlaneState and NodeSpecer

This commit removes usages of NodeSpecer and ControlPlaneState from the
BGP Control Plane.

With the inclusion of LocalNodeStore we no longer need the NodeSpecer
abstraction nor the ControlPlaneState.

Moving forward all ConfigReconcilers will get a read-only node.LocalNode object
to inspect the current state of the Cilium node.

If a ConfigReconciler needs more information then this they can now
"request" depedencies via the Hive/Cell depedency injection graph (DIG).

All ConfigReconcilers are part of the DIG and thus can declare
depedencies via their constructor's parameters.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
bgpv1: remove NodeSpecer and ControlPlaneState references

This commit removes all references to NodeSpecer and ControlPlaneState.

The previous commit removed them from the BGP CP.

See the previous commit for further explanation.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
bpgv1,tests: make tests work with NodeSpecer and ControlPlaneState omit

This commit updates the BGP-CP unit tests to work correct after the
removal of NodeSpecer and ControlPlaneState structures.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -352,46 +353,9 @@ func (c *Controller) Reconcile(ctx context.Context) error {
 		return fmt.Errorf("invalid BGP peering policy %s: %w", policy.Name, err)
 	}
 
-	// parse any virtual router specific attributes defined on this node via
```
This whole block goes away, since all this information can be gained just from looking at 'localNode'

rastislavs commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -217,7 +172,12 @@ func (c *Controller) Run(ctx context.Context) {
+
+	// setup a reconciliation trigger on LocalNodeStore changes
+	c.LocalNodeStore.Observe(ctx, func(node node.LocalNode) { c.Sig.Event(struct{}{}) }, func(err error) {
+		l.WithError(err).Info("LocalNodeStore observe has yielded. Reconciliation with no longer be triggered for LocalNode changes")
```
```suggestion
		l.WithError(err).Info("LocalNodeStore observe has yielded. Reconciliation will no longer be triggered for LocalNode changes")
```

ldelossa replied:
fixed

rastislavs commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -201,18 +202,26 @@ func (m *BGPRouterManager) registerBGPServer(ctx context.Context, c *v2alpha1api
-		return err
+		nodeIP := node.GetNodeIP(false)
+		if nodeIP.IsUnspecified() {
+			return fmt.Errorf("failed to resolve router id")
```
Nit: maybe wrap the error returned from `ResolveRouterID` here?

ldelossa replied:
fixed

rastislavs commented on pkg/bgpv1/agent/annotations.go:
```diff
@@ -103,6 +104,15 @@ func (e ErrMulti) Error() string {
+	if _, ok := a[localASN]; ok {
+		if parsed, err := netip.ParseAddr(a[localASN].RouterID); err == nil && !parsed.IsUnspecified() {
+			return parsed.String(), nil
+		}
```
Nit: Would be nicer to return a wrapped error from `ParseAddr` if it returns an error, so that in case that incorrect address is provided in the annotation, you would get a parse error instead of "router id not specified by annotation"

ldelossa replied:
fixed

rastislavs reviewed (commented):
found just some nits, overall a great change!

harsimran-pabla commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -217,7 +213,12 @@ func (c *Controller) Run(ctx context.Context) {
+	l.Info("Starting LocalNodeStore Observer")
+
+	// setup a reconciliation trigger on LocalNodeStore changes
+	c.LocalNodeStore.Observe(ctx, func(node node.LocalNode) { c.Sig.Event(struct{}{}) }, func(err error) {
```
Should this be part of Controller.Start,? 

Since policy-observer is defined there, it would also make sense to have node-observer in it.

ldelossa replied:
I'm not sure it matters, as long as we are observing before we enter the `run` loop. 

But maybe it's best to do it there for consistency?

harsimran-pabla replied:
Yeah, it was mostly for consistency since Observe does not need to be started in its own goroutine.

ldelossa replied:
@harsimran-pabla - I attempted this change, but it seems to have broken the Observer. I no longer see events if I place it in the Start hook. 

I'm inclined to leave it as is for now and Ill investigate it a bit further once a separate PR

harsimran-pabla replied:
ack, no issues.

harsimran-pabla commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -217,7 +213,12 @@ func (c *Controller) Run(ctx context.Context) {
+
+	// setup a reconciliation trigger on LocalNodeStore changes
+	c.LocalNodeStore.Observe(ctx, func(node node.LocalNode) { c.Sig.Event(struct{}{}) }, func(err error) {
+		l.WithError(err).Info("LocalNodeStore observe has yielded. Reconciliation with no longer be triggered for LocalNode changes")
```
Should this be error log ? `l.WithError(err).Info -> l.WithError(err).Error`

ldelossa replied:
Well, we get an error, but most likely, the system stopped the observer for some valid reason, which is probably not an error?

harsimran-pabla commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -201,18 +202,26 @@ func (m *BGPRouterManager) registerBGPServer(ctx context.Context, c *v2alpha1api
+		nodeIP := node.GetNodeIP(false)
+		if nodeIP.IsUnspecified() {
+			return fmt.Errorf("failed to resolve router id: %w", err)
+		}
```
Missing assignment of nodeIP to routerID 
225: `routerID = nodeIP.String()`

harsimran-pabla commented on pkg/bgpv1/manager/reconcile.go:
```diff
@@ -100,78 +92,86 @@ func preflightReconciler(
+		nodeIP := p.Node.GetNodeIP(false)
+		if nodeIP.IsUnspecified() {
+			return fmt.Errorf("failed to resolve router id")
+		}
```
Missing routerID = nodeIP.String()

harsimran-pabla commented on pkg/bgpv1/agent/annotations.go:
```diff
@@ -103,6 +104,15 @@ func (e ErrMulti) Error() string {
+func (a AnnotationMap) ResolveRouterID(localASN int64) (string, error) {
+	if _, ok := a[localASN]; ok {
+		if parsed, err := netip.ParseAddr(a[localASN].RouterID); err == nil && !parsed.IsUnspecified() {
+			return parsed.String(), fmt.Errorf("failed to parse RouterID for local ASN %v: %w", localASN, err)
```
Something is not right here, on err == nil, we should not be returning err.

harsimran-pabla commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -217,7 +172,12 @@ func (c *Controller) Run(ctx context.Context) {
+	l.Info("Starting LocalNodeStore Observer")
+
+	// setup a reconciliation trigger on LocalNodeStore changes
+	c.LocalNodeStore.Observe(ctx, func(node node.LocalNode) { c.Sig.Event(struct{}{}) }, func(err error) {
```
General observation, observe is not triggering event on changes to annotations in node object.

ldelossa replied:
You're right, issue is here: https://github.com/ldelossa/cilium/blob/ldelossa/bgpv1-configreconciler-hivecell/pkg/endpointmanager/host.go#L70

We do not trigger any update to the LocalNodeStore on k8s node updates.

ldelossa replied:
Fixed with latest commit.

ldelossa committed:
localnode: consider annotations in LocalNodeStore updates

Prior to this PR only label changes made to the Kubernetes Node object
would propagate to LocalNodeStore updates.

This commit adds Annotations to this mix as well.

This is useful as the BGP-CP now observes the LocalNodeStore and
requires events to occur when annotations are updated.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

YutaroHayakawa reviewed (approved):
LGTM overall 👍 Thanks for your big effort! This makes code much simpler and easy to change.

One non-blocking suggestion. How about splitting each reconcilers into separated files? That makes the code easy to read, reduce the risk of conflict when we introduce a new reconciler, etc.

rastislavs reviewed (approved):
LGTM

harsimran-pabla commented on pkg/endpointmanager/host.go:
```diff
@@ -61,17 +63,37 @@ func (mgr *endpointManager) OnUpdateNode(oldNode, newNode *slim_corev1.Node,
+	annotationsEqual := comparator.MapStringEquals(oldNodeAnnotations, newNodeAnnotations)
+
+	// if labels have changed we need to recompute security ID.
+	if labelsEqual {
```
Should this be `if !labelsEqual` ?

ldelossa replied:
Yup! Good call.

aditighag commented on pkg/endpointmanager/host.go:
```diff
@@ -61,17 +63,37 @@ func (mgr *endpointManager) OnUpdateNode(oldNode, newNode *slim_corev1.Node,
+	newNodeAnnotations := newNode.GetAnnotations()
 
-	node.SetLabels(newNodeLabels)
+	labelsEqual := comparator.MapStringEquals(oldNodeLabels, newNodeLabels)
+	annotationsEqual := comparator.MapStringEquals(oldNodeAnnotations, newNodeAnnotations)
```
There is a factory function available to compare annotations - https://github.com/cilium/cilium/blob/main/pkg/k8s/factory_functions.go#L67. Any reason why we don't use it here? The watcher subsystem uses the same function.

ldelossa replied:
Good call, no reason here, I just stumbled upon the `comparator` pkg in the Watcher code before I actually saw that function.

ldelossa replied:
Looking at this again, I think I do prefer to use the `comparator` package here. 

1. There is no comparable `LabelsEqual` functions, so I'd be importing `comparator` anyway.
2. I don't need to filter the annotations for comparison.
3. The watcher package also uses `comparator` internally as well.

aditighag replied:
Well, there is a difference in the implementation. The factory function checks if a specific annotation is equal or not. Maybe you need MapStringEquals here for comparing all annotations no a node.

ldelossa replied:
Yes I think so, at this level, we simply want to know if any annotation has changed between the old and the new node.

aditighag commented on pkg/endpointmanager/host.go:
```diff
@@ -61,17 +63,37 @@ func (mgr *endpointManager) OnUpdateNode(oldNode, newNode *slim_corev1.Node,
+	// Perform a SetMultiAttributes if both labels and annotations has changed,
+	// since these Set operations will wake up Observers on LocalNodeStore its
+	// better to perform one update if multiple attributes change.
+	if !labelsEqual && !annotationsEqual {
+		node.SetMultiAttributes(func(n *node.LocalNode) {
+			n.Labels = newNodeLabels
+			n.Annotations = newNodeAnnotations
+		})
+	} else if !labelsEqual {
+		node.SetLabels(newNodeLabels)
+	} else if !annotationsEqual {
+		node.SetAnnotations(newNodeAnnotations)
```
Can we pass the `labelsEqual` and `annotationsEqual` flags to `SetMultiAttribues`, and fold the if-else conditions?

ldelossa replied:
That would make SetAnnotations pretty "specific", no? Right now the method doesn't care 'why' you want to set multiple attributes, its just an interface to do so. This seems appropriate to me? Since it may be used in a completely different setting for a completely different evaluation of Node fields.

aditighag replied:
If `MultiSetAttributes` were used in multiple places, that might've made sense. But only two attributes are being set here depending on the labels and annotations equal flags. Anyway it's your call.

ldelossa replied:
Yes, I do think we *need* an opinionated way to set multiple values on LocalNode without causing LocalNodeStore subscribers to wake up multiple times.

aditighag commented on pkg/k8s/watchers/node.go:
```diff
@@ -34,7 +34,8 @@ type NodeUpdate interface {
 func nodeEventsAreEqual(oldNode, newNode *slim_corev1.Node) bool {
-	return comparator.MapStringEquals(oldNode.GetLabels(), newNode.GetLabels())
+	return comparator.MapStringEquals(oldNode.GetLabels(), newNode.GetLabels()) &&
+		comparator.MapStringEquals(oldNode.GetAnnotations(), newNode.GetAnnotations())
```
`nodeEventsAreEqual` is being called in a generic event loop that tracks changes to the node ep, and notifies the subscribers of node events. Your commit message only talks about the annotations update being tracked by the BGP subsystem, so how do we know that the updated function doesn't cause any regressions for other subsystems that are listening to node events?

ldelossa replied:
My presumed consequence of this change is that all existing code works as normal, but may now also receive an event when the k8s Node object associated with the current Cilium agent has its annotations changed. 

I'm not capable to testing every sub-system, but I do rely on tests (which are all passing) to obtain confidence in this change. 

I felt it was safe to assume that sub-systems that are listening for NodeEvents would assume the same semantics as a change to the k8s Node object itself which fires for **any** change? 

Do you feel there maybe an issue for a sub-system to receive a NodeEvent on both labels and annotations changes?

aditighag replied:
>I'm not capable to testing every sub-system, but I do rely on tests (which are all passing) to obtain confidence in this change.

Yes, that's fair. But the PR doesn't seem to add a test that updates node annotations, so how the other subsystems behave when they receive additional events is unknown?

>Do you feel there maybe an issue for a sub-system to receive a NodeEvent on both labels and annotations changes?

I can't say definitively, but I looked at the service cache subscriber, and it seems to assume that updates only pertain to labels - https://github.com/cilium/cilium/blob/main/pkg/k8s/service_cache.go#L739-L739. 

I think it's worthwhile to bring this up in the community meeting, or in the public Slack channel (maybe just as a heads-up). Wdyt?

ldelossa replied:
Yup I do agree. Our sig was actually bringing up your point that some sub-systems may think an update **implicitly** indicates that only labels are updated and whether this is an issue or not. 

Bringing it up in next comm meeting sounds like a good idea and I'll add a message in public Slack before hand.

ldelossa replied:
@aditighag I went through each implementation of the `subscriber.Node` interface to confirm that an event for annotations will not cause an issue. 

Following Chris's response in this thread: https://cilium.slack.com/archives/C2B917YHE/p1692050432666799, I think this change is rather safe. 

I do have a few other questions around our node-syncing design, but they are a bit orthogonal to this PR, and will be doing some investigating further.

aditighag replied:
Thanks for checking! It would be nice to have a test that mutates node annotations: something to keep in mind for follow-ups.

ldelossa replied:
I'm def not opposed to this, but I am having a bit of trouble understanding where that test would live and what is under test in that scenario. 

Is it, testing each sub-system can handle an event where a node annotation is updated or are you more thinking a test within the watcher code itself?

aditighag replied:
> testing each sub-system can handle an event where a node annotation is updated 

^^ Yes, mainly this. Maybe the control plane test framework is well suited to test such a functionality?

aditighag reviewed (changes requested):
Thanks for highlighting the commits that are not BGP specific. I've left some review comments. Mainly, it's not clear to me how some of the changes may affect other subsystems subscribing to node events.
